### DISCUSSION TITLE
Missing javascript client version

### DISCUSSION THREAD

**Discussion by p4tr1ck213** [Azure/azure-mobile-apps#661] [Q&A] [answered]
First of all thanks for this repository which helps us a lot in our planning strategy on azure webapps. Is it planned to offer a javascript/typescript version for the client side, so that offline PWAs with e.g. an asp.net angular combination are also possible? Or what would be the right approach?

**Comment by adrianhall** [comment, marked as answer]
It's relatively easy to write a JS/TS version of the remote client, but much harder to write the equivalent for an offline client.  This is because the on-device database options are not suitable for offline capabilities.

As a result, I don't have a plan to release an npm package for the client in JS/TS.

**Reply by p4tr1ck213** [reply]
Thank you for your quick response. Isn't it a good idea to use webassemblies of the clientcode part that then uses js indexdb for the offline store?

**Reply by adrianhall** [reply]
Not really.  What happens is that the entire SQLite binary database is stored as a blob and then loaded into memory.  For larger offline stores, this results in a long load time - in a system notorious for long load times.  The unfortunate thing is "larger" is actually relatively small datasets.

A better solution is to provide an IndexedDb offline store provider.  However, the current code structure doesn't allow for that, which means the existing code is completely non-suited for the capabilities required.  That leads me to conclude that I need to do a major rewrite of the codebase (which you can find mostly in the archive branch - only the LINQ compatibility layer is missing).  Since I am still working on the .NET codebase and no-one has stepped up to assist, that leaves me with no choice but to de-prioritize the effort.

**Reply by p4tr1ck213** [reply]
To write an exrtra idb provider also seems to me the best solution. It's a pity that you don't have more help from ms/azure colleagues with this important project. Many other providers support offline PWA solutions and I would really like to use Azure MBaaS in the future because it fits best into the .Net and Azure ecosystem and apart from this little thing has everything I need.
